lib/backend/watchersyncer/watchersyncer.go

```diff
@@ -0,0 +1,200 @@
+	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
+)
+
+const (
+	maxUpdatesToConsolidate = 1000
+)
+
+// ResourceType groups together the watch and conversion information for a
+// specific resource type.
+type ResourceType struct {
+	// The ListInterface used to perform a watch on this resource type.
+	ListInterface model.ListInterface
+
+	// An update processor used to convert the update event prior to it being sent in
+	// the Syncer.
+	UpdateProcessor SyncerUpdateProcessor
+}
+
+// SyncerUpdateProcessor is used to convert a Watch update into one or more additional
+// Syncer updates.
+type SyncerUpdateProcessor interface {
+	// Process is called to process a watch update.  The processor may convert this
+	// to zero or more updates.  The processor may use these calls to maintain a local cache
+	// if required.  It is safe for the processor to send multiple duplicate adds or deletes
+	// since the WatcherSyncer maintains it's own cache and will swallow duplicates.
+	// A KVPair with a nil value indicates a delete.  A non nil value indicates an add/modified.
+	// The processor may respond with any number of adds or deletes.
+	Process(*model.KVPair) ([]*model.KVPair, error)
+
+	// SyncStarting is called when syncer is starting a full sync for the associated resource
+	// type.  That means it is first going to list current resources and then watch for any updates.
+	// If the processor maintains a private internal cache, then the cache should be cleared at
+	// this point since the cache will be re-populated from the sync.
+	SyncStarting()
+}
+
+// New creates a new multiple Watcher-backed api.Syncer.
+func New(client api.Client, resourceTypes []ResourceType, callbacks api.SyncerCallbacks) api.Syncer {
+	rs := &watcherSyncer{
+		watcherCaches: make([]*watcherCache, len(resourceTypes)),
+		results:       make(chan interface{}, 2000),
+		callbacks:     callbacks,
+	}
+	for i, r := range resourceTypes {
+		rs.watcherCaches[i] = newWatcherCache(client, r, rs.results)
+	}
+	return rs
+}
+
+// watcherSyncer implements the api.Syncer interface.
+type watcherSyncer struct {
+	status        api.SyncStatus
+	watcherCaches []*watcherCache
+	results       chan interface{}
+	numSynced     int
+	callbacks     api.SyncerCallbacks
+}
+
+func (rs *watcherSyncer) Start() {
+	log.Info("Start called")
+	go rs.run()
+}
+
+// Send a status update and store the status.
+func (rs *watcherSyncer) sendStatusUpdate(status api.SyncStatus) {
+	log.WithField("Status", status).Info("Sending status update")
+	rs.callbacks.OnStatusUpdated(status)
+	rs.status = status
+}
+
+// run implements the main syncer loop that loops forever receiving watch events and translating
+// to syncer updates.
+func (rs *watcherSyncer) run() {
+	log.Info("Running - sending status update")
+	rs.sendStatusUpdate(api.WaitForDatastore)
+	log.Info("Starting watcher caches")
+	for _, wc := range rs.watcherCaches {
+		go wc.run()
+	}
+
+	log.Info("Starting main watch processing loop")
```

Comment:
"Starting main event processing loop" would be more accurate - the events might come from a list or a watch.